File: 3.0/python/32.py
from sys import setrecursionlimit
setrecursionlimit(100000)

n, m = map(int, input().split())
graph = [[] for _ in range(n+1)]
visited = {i: False for i in range(len(graph))}
components = {}

for _ in range(m):
    v, e = map(int, input().split())
    graph[v].append(e)
    graph[e].append(v)

def dfs(graph, visited, now, compc):
    visited[now] = True

    if components.get(compc) == None:
        components[compc] = set()
    
    if len(graph[now]) > 0:
        components[compc].update(graph[now])
    else:
        components[compc].add(now)

    for neig in graph[now]:
        if not visited[neig]:
            dfs(graph, visited, neig, compc)

compc = 1
for i in range(1, n+1):
    if not visited[i]:
        dfs(graph, visited, i, compc)
        compc += 1
        
print(compc-1)
for k, v in components.items():
    print(len(v))
    print(*v)

File: 3.0/golang/32.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var components = make(map[int]map[int]struct{})

func dfs(graph [][]int, visited map[int]bool, now, compc int) {
	visited[now] = true

	if _, ok := components[compc]; ok == false {
		components[compc] = map[int]struct{}{}
	}

	if len(graph[now]) > 0 {
		for _, v := range graph[now] {
			components[compc][v] = struct{}{}
		}
	} else {
		components[compc][now] = struct{}{}
	}

	for _, neig := range graph[now] {
		if visited[neig] == false {
			dfs(graph, visited, neig, compc)
		}
	}
}

func main() {
	var N, M int
	in := bufio.NewReader(os.Stdin)
	fmt.Fscanf(in, "%d %d\n", &N, &M)

	graph := make([][]int, N+1)
	visited := make(map[int]bool)
	for i := 0; i < N+1; i++ {
		visited[i] = false
	}

	var v, e int
	for i := 0; i < M; i++ {
		fmt.Fscanf(in, "%d %d\n", &v, &e)
		graph[v] = append(graph[v], e)
		graph[e] = append(graph[e], v)
	}

	compc := 0
	for i := 1; i < N+1; i++ {
		if visited[i] == false {
			compc++
			dfs(graph, visited, i, compc)
		}
	}

	fmt.Println(compc)
	for _, v := range components {
		fmt.Println(len(v))
		for k := range v {
			fmt.Print(k, " ")
		}
		fmt.Print("\n")
	}
}
